### ISSUE TITLE
[NEXT]Convert bootcamp cost calculator to react

### ISSUE THREAD

**Issue by Bouncey** [freeCodeCamp/freeCodeCamp#18094]
[The coding bootcamp cost calculator](https://www.freecodecamp.org/coding-bootcamp-cost-calculator) needs to be converted to a react component.

A PR addressing this should be opened against the `next` branch

**Comment by artoodeeto** [comment]
@huyenltnguyen hello. anyway i could help? its my first time to make contributions though.

**Comment by huyenltnguyen** [comment]
Hi @bouncey, I just took a look at the code and I would really appreciate some pointers from you.
- Is it just me or the app is not functioning right now? And if so, I suppose the task here is to convert only the UI to React, without implementing the logic? I would love to help fix/complete the app, btw, if I could get some user stories.
- Where is the best place to save the newly created JS file? In the same directory as the jade file maybe?
- I'm not familiar with React on server side, so I'm wondering how I can configure the router to render the file for reviewing purpose. Or should I just temporarily build the app somewhere else then copy & paste it in once I'm happy with it? That could help avoid the headache.

Thank you in advance.

---

Hi @artoodeeto, thank you for offering to help! I just gave it a quick look but I don't think I can start working on it at the moment (probably in a few days). Feel free to take it from me, but let's wait for some guidance from Bouncey. If you have successfully got the repo up and running, then just poke around in the meantime maybe, otherwise let me know if I can be of any help :)

**Comment by Bouncey** [comment]
@huyenltnguyen @artoodeeto Sorry, I should have been more descriptive as we are converting the client app to `gatsby`

You should be able to get the development environment working by running `npm run develop` from the project root, ensuring that you have mongo running locally

This issue requires:
1. A new file created at `/client/src/pages/coding-bootcamp-cost-calculator.js`
1. `coding-bootcamp-cost-calculator.js` should export a react component as default (any other file in the `/pages` directory could be used for guidence)
1. All `jQuery` code in `/api-server/public/js/calculator.js` should be refactored into the react component (or where it makes sense to have it) without a dependancy on `jQuery`. Use of component state should be enough for this.

**Comment by huyenltnguyen** [comment]
@Bouncey Awesome! Things are much clearer now. Thanks a lot!

---
@artoodeeto 
> can we do this together?

Totally yes! The reason I offered it to you is because I was afraid I would slow you down. Now that is going to be a sure thing since we are going to collaborate. I will do my best though 😄 
> just tell me your schedule and i can adjust mine. i just need to know how to start this.

No need to adjust your schedule to fit mine. I would say let's break this up into small pieces. Each of us works on half of them and reviews the other person's work. This will be the first time I use Git to collaborate, so it will be a learning experience for me, too.

I saw you in the Contributors chatroom, is it okay if I reach you via Gitter?

**Comment by huyenltnguyen** [comment]
Hi @Bouncey, I have a quick question. So we need to create a `coding-bootcamp-cost-calculator.js` file in `/client/src/pages/`, but I'm not sure how to review the component. Do we need to create a route for it?

**Comment by Bouncey** [comment]
By adding the `coding-bootcamp-cost-calculator` file in `src/pages` and exporting a component from it, Gatsby will create a route at `http://localhost:8000/coding-bootcamp-cost-calculator`

Hope this helps

**Comment by huyenltnguyen** [comment]
@Bouncey It does help, but now I'm annoyed that I have no clue how it works. I already put Gatsby-the-black-magic-thing on my to-learn list. I will definitely have to find time to learn it 😄 

Many thanks.

**Comment by huyenltnguyen** [comment]
Hi @junaidmanzur, thanks for offering the help.

@artoodeeto  and I are slowly making progress on this. We are almost done with the UI and will soon try implementing the logic for the app. I was planning to post an update here when we complete the UI. I'm sorry for not actively reporting on our progress.

**Comment by huyenltnguyen** [comment]
Hi @Bouncey, it has been awhile and I sincerely apologize for not actively keeping you posted on the progress.

@artoodeeto and I have completed the UI of the app. Both of us have been on a very tight schedule, so this is a great milestone for us.

We have started reading the `/api-server/public/js/calculator.js` file and trying to figure out what does what. However, the code is quite complicated to understand and the live app is not working, so the challenge level is doubled for us.


I did some inspection on the live app and found that it's because the app doesn't load or fails to load `jQuery` and `d3`. I did not bother to try to fix it since I assume it will be removed after the React conversion. Instead, I extracted the app and did some tweaking to make it work, then posted it on [codepen](https://codepen.io/huyenltnguyen/full/zJXdzB/) so that we can play with it and see how things work.

We would like to ask for some guidance on these things:
- Since we want to convert the jQuery app to React, I assume we would like to use a React-powered `d3`  package as a substitution for `d3`? Could suggest a package that we can use? I'm leaning toward [react-d3](https://github.com/esbullington/react-d3) but still want to seek your input.
- There is a [AJAX call](https://github.com/freeCodeCamp/freeCodeCamp/blob/5fedefb74f1f5e0440f5dad8956a7cba8009971c/api-server/public/js/calculator.js#L3) that doesn't work on local. We are just using `import` in development for now, but should we change it to something else or just leave it as it is?

Thank you.

---
This app is quite heavy on `d3` but neither @artoodeeto  nor I am familiar with `d3`. My approach is to just replace each `d3` API in the app with its correspoding `react-d3` API, and learn `d3` in the process.

We understand that you guys are working super hard to get the `next` branch merged to `master` in the very near future. We don't want to slow you down, so if you guys would like to have this task completed sooner, we are happy to pass it to more available contributors and assist them to the best of our ability. Otherwise, please rest assured that we are still trying to do our best to finish the app asap.

**Comment by artoodeeto** [comment]
@Bouncey @huyenltnguyen i would like to apologies for being MIA. I was so busy with school. i apologies. hope youll understand. i hope i can and will still be allowed in the future. thank you and my bad if this got delayed.

**Comment by ertankara** [comment]
if this task is available I would like to take this one.

**Comment by huyenltnguyen** [comment]
Hi @ertankara, I'm still working on this task, but the progress is super slow due to my busy schedule :disappointed: I still want to continue the task, but I need some help if we want to speed it up. I'm happy to pass it to more available contributors though. Are you on [Gitter](https://gitter.im) by any chance? If so, please send me a direct message at gitter.im/huyenltnguyen so we can discuss further.

@artoodeeto All good. No need to apologize, my friend. Let me know if you can find some free time, I may need your help with this app. Otherwise, come back and contribute when you're more available. fCC always welcomes contributors, so don't worry 😄

**Comment by artoodeeto** [comment]
@huyenltnguyen I am back 100%. can you run me through the project again? I can learn D3. I was planning to learn it before now is the right time.

**Comment by byronluk** [comment]
Hi I'd like to help out if its needed. Lmk @huyenltnguyen @artoodeeto , thanks!

**Comment by abavish** [comment]
Hello, I'd love to help too! Please let me know if I could assist in any way.

**Comment by artoodeeto** [comment]
yes!! @byronluk @abavish im gonna add you on gitter. we created a group. yey we can finish it now. @huyenltnguyen 

i added you on gitter guys. :)

**Comment by abavish** [comment]
Thank you @artoodeeto! :)

**Comment by QuincyLarson** [comment]
Closing in favor of doing this #34494
